Galaxy Webhooks

bgruening opened the pull request:
# Objective

Galaxy has a rich set of plugin systems for visualizations, interactive environments, toolbox filters, tours and many more. What is missing currently is to extend the Galaxy UI (the client) with small code snippets like additional icons, menu entries or other useful information that an admin wants to display to the user.

The first time @erasche and I wanted to have such a system was during the development of IEs, in order to include an optional entry in the workflow menu. The second time such a system would have been handy was during the development of the chat-server which was included in the last release (#2515). Such a functionality would be better as a plugin and not as core component.

This PR solves this shortcoming by introducing `webhooks` into Galaxy - a system which can be used to write small JS and/or Python functions to change predefined locations in the Galaxy client.

For now we have 4 of such predefined locations, which we call entry points.
- tool (after tool execution)
- workflow (after workflow execution)
- masthead (at the top level masthead)
- history-menu (inside History Panel menu)

For all four entry point we have developed example plugins to demonstrate the capabilities of the proposed system. For example after a tool or workflow was executed we can show either XKCD comics or PhD comics ...

![](https://github.com/bgruening/galaxy/blob/6f8deb6f4a07bd57493d24cd90a9862dbf7e6d0c/doc/source/admin/special_topics/images_webhooks/tool.png?raw=true)

For the masthead we have two example plugins, one is just showing an icon which links out to biostars and the other displays the name of the user. The latter is just for demonstrating that a plugin author has access to the user object and also to the history which give you a lot of flexibility in webhook development.

![](https://github.com/bgruening/galaxy/blob/6f8deb6f4a07bd57493d24cd90a9862dbf7e6d0c/doc/source/admin/special_topics/images_webhooks/masthead.png?raw=true)

The history menu can also be extended as shown in the following image. It has currently no functionality, but if this PR gets accepted we plan to create one for launching IEs. 

![](https://github.com/bgruening/galaxy/blob/6f8deb6f4a07bd57493d24cd90a9862dbf7e6d0c/doc/source/admin/special_topics/images_webhooks/history-menu.png?raw=true)

# Plugin activation

Webhooks are deactivated by default. They can be activated by making sure that the path in set in your `galaxy.ini` and that you have activated all plugins that you want to use manually by changing the config of each webhook:

```
config/plugins/webhooks/{TOOL_NAME}/config/{TOOL_NAME}.yml
```

and set the parameter activate to true:

```
activate: true
```
# The future

We have a lot of ideas and a few webhooks already in development which we would like to propose for inclusion (or we will create a different github repo for them.)
- RSS reader
- Tool run-time estimation
- PubMed integration taking the current worklfow into consideration and propose new manuscritps
- recommendation system for tools (no clue if this works, but we will try)
- communication server as webhook

I really hope we can trigger some cool, new (but optional) features with this and make the Galaxy more fun and useful for our users. Gamification of Galaxy should be possible with this and I can tell you these comics makes it already a lot more funny to analyse data for our users :)

One thing we are working on in an other branch is to extend the user-preference panel in a way that allows every user to activate deactivate webhooks. But this is for another PR.
## Documentation

More information can be found in our documentation which is included in this PR: https://github.com/bgruening/galaxy/blob/feature/plugin-system/doc/source/admin/webhooks.rst
# Thanks!

With this I want to thank our awesome intern Evgeny Anatskiy (@anatskiy) who has implemented this system during the last weeks, Joachim Wolff (@joachimwolff) for testing and very useful comments and Eric Rasche (@erasche), Jeremy Goecks (@jgoecks) and Aysam Guerler (@guerler) for inspiring discussion during a GCC beer night.

dannon committed:
Merge branch 'release_16.01'

jmchilton committed:
Merge branch 'release_16.04'

dannon committed:
Merge branch 'release_16.04'

dannon committed:
Merge branch 'release_16.04'

dannon committed:
Merge branch 'release_16.04'

dannon committed:
Merge branch 'release_16.04'

dannon committed:
Merge branch 'release_16.04'

dannon committed:
Merge branch 'release_16.04'

anatskiy committed:
add Webhooks draft

anatskiy committed:
add Webhooks draft

anatskiy committed:
add Webhooks draft

anatskiy committed:
add Webhooks draft

anatskiy committed:
add Webhooks draft

anatskiy committed:
add Webhooks draft

anatskiy committed:
add Webhooks draft

remove extra files

anatskiy committed:
add Webhooks draft

anatskiy committed:
add Webhooks draft

anatskiy committed:
remove extra files

anatskiy committed:
remove extra files

anatskiy committed:
remove extra files

anatskiy committed:
remove extra files

anatskiy committed:
remove extra files

anatskiy committed:
remove extra files

anatskiy committed:
remove extra files

anatskiy committed:
remove extra files

anatskiy committed:
Merge branch 'feature/plugin-system' of github.com:bgruening/galaxy into feature/plugin-system

dannon committed:
Merge branch 'release_16.04'

dannon committed:
Merge branch 'release_16.04'

dannon committed:
Merge branch 'release_16.04'

anatskiy committed:
add very basic plugin system (for tools)

anatskiy committed:
add very basic plugin system (for tools)

anatskiy committed:
add very basic plugin system (for tools)

anatskiy committed:
add very basic plugin system (for tools)

anatskiy committed:
add very basic plugin system (for tools)

anatskiy committed:
add very basic plugin system (for tools)

anatskiy committed:
add very basic plugin system (for tools)

anatskiy committed:
add very basic plugin system (for tools)

anatskiy committed:
add first webhook

anatskiy committed:
add first webhook

anatskiy committed:
add first webhook

anatskiy committed:
add first webhook

anatskiy committed:
add first webhook

anatskiy committed:
add first webhook

anatskiy committed:
add first webhook

anatskiy committed:
add first webhook

anatskiy committed:
add one more plugin

anatskiy committed:
add one more plugin

anatskiy committed:
add one more plugin

anatskiy committed:
add one more plugin

anatskiy committed:
add one more plugin

anatskiy committed:
add one more plugin

anatskiy committed:
add one more plugin

anatskiy committed:
add one more plugin

anatskiy committed:
topbar webhooks testing

anatskiy committed:
topbar webhooks testing

anatskiy committed:
topbar webhooks testing

anatskiy committed:
topbar webhooks testing

anatskiy committed:
topbar webhooks testing

anatskiy committed:
topbar webhooks testing

anatskiy committed:
topbar webhooks testing

anatskiy committed:
topbar webhooks testing

anatskiy committed:
Merge branch 'feature/plugin-system' of github.com:bgruening/galaxy into feature/plugin-system

anatskiy committed:
renamings and fixes

anatskiy committed:
renamings and fixes

anatskiy committed:
renamings and fixes

anatskiy committed:
renamings and fixes

jmchilton committed:
Merge branch 'release_16.07'

anatskiy committed:
- unify webhooks entry point
- simlify webhooks code (on load)
- improve code readability
- clean up

anatskiy committed:
- unify webhooks entry point
- simlify webhooks code (on load)
- improve code readability
- clean up

anatskiy committed:
- unify webhooks entry point
- simlify webhooks code (on load)
- improve code readability
- clean up

anatskiy committed:
- add new entry point (Workflow)
- temporarily hide masthead webhook (due to instability)

anatskiy committed:
- add new entry point (Workflow)
- temporarily hide masthead webhook (due to instability)

anatskiy committed:
- add new entry point (Workflow)
- temporarily hide masthead webhook (due to instability)

anatskiy committed:
- add new entry point (Workflow)
- temporarily hide masthead webhook (due to instability)

dannon committed:
Merge branch 'release_16.07'

anatskiy committed:
add new entry point for History Panel menu

anatskiy committed:
add new entry point for History Panel menu

anatskiy committed:
add new entry point for History Panel menu

anatskiy committed:
bug fixes and improvements

anatskiy committed:
bug fixes and improvements

anatskiy committed:
bug fixes and improvements

anatskiy committed:
bug fixes and improvements

anatskiy committed:
Merge branch 'feature/plugin-system' of github.com:bgruening/galaxy into feature/plugin-system

# Conflicts:
#	client/galaxy/scripts/layout/menu.js
#	client/galaxy/scripts/mvc/tool/tool-form.js
#	static/maps/layout/menu.js.map
#	static/maps/mvc/tool/tool-form.js.map
#	static/scripts/bundled/analysis.bundled.js
#	static/scripts/bundled/analysis.bundled.js.map
#	static/scripts/bundled/libs.bundled.js
#	static/scripts/bundled/libs.bundled.js.map
#	static/scripts/layout/menu.js
#	static/scripts/mvc/tool/tool-form.js

anatskiy committed:
fix bugs after previous commit

anatskiy committed:
fix bugs after previous commit

anatskiy committed:
pass trans object into the helper function

anatskiy committed:
pass trans object into the helper function

bgruening committed:
Add initial docs

bgruening committed:
Add initial docs

anatskiy committed:
- show Webhook on a workflow page
- fix merge errors

anatskiy committed:
- merge

anatskiy committed:
make client

anatskiy committed:
- remove separate functions for adding masthead/history-menu plugins
- choose a random tool/workflow plugin only among active ones

anatskiy committed:
bug fix

anatskiy committed:
tiny fixes

bgruening committed:
Include a section about plugin dependencies.

joachimwolff committed:
Extended the documentation of webhooks and added an images directory.

joachimwolff committed:
Added images directory for webhooks documentation

joachimwolff committed:
Fixed formatting errors

joachimwolff committed:
Fixed formatting errors

joachimwolff committed:
Fixing errors

joachimwolff committed:
Fixing errors

joachimwolff committed:
Fixing errors

joachimwolff committed:
Fixing errors

joachimwolff committed:
Fixing errors

joachimwolff committed:
Fixing errors

joachimwolff committed:
Fixing errors

bgruening committed:
Merge pull request #21 from joachimwolff/feature/plugin-system

Feature/plugin system

anatskiy committed:
added missing dependencies check

anatskiy committed:
Update webhooks.rst

escape missing dependencies

anatskiy committed:
Update webhooks.rst

bgruening committed:
make activation procedure more clear

gregvonkuster commented:
Wow!  So cool!  ;)

guerler commented on client/galaxy/scripts/layout/menu.js:
```diff
@@ -1,396 +1,418 @@
+                            if (webhook.activate) {
+                                Galaxy.page.masthead.collection.add({
+                                    id      : webhook.name,
+                                    icon    : (typeof webhook.config.icon != 'undefined') ? webhook.config.icon : '',
```
Minor, but this can be written as: `icon: webhook.config.icon`, right?

jmchilton commented on templates/webapps/galaxy/workflow/run_complete.mako:
```diff
@@ -0,0 +1,55 @@
+<%inherit file="/base.mako"/>
```
This file has been removed from `dev` - should it be removed here as well. Perhaps this needs a merge with latest dev.

bgruening replied:
I merged latest dev into this branch and it is still here.
We modified this file at a time were `run_complete.mako` still existed but it should not be in this branch anymore :)
Any help appreciated! I fear a rebase as so many JS changes are in that a rebase is painful.

dannon replied:
(merge with -dev didn't resolve this -- opening separate PR to strip it out (again))

jmchilton commented:
First of two notes - this one with requests for changes related to configuration:
- [ ] These operate different from IEs and Vizs that just require the plugin be present. Can we adjust the behavior to just load it if it is there? Maybe create a demo plugin directory and real plugin directory (the latter being empty) - and just update the documentation to state that the demo plugins can be enabled by adding that directory to the config.
- [ ] Can we update the config option to allow multiple plugin directories. Tours, viz, and IEs allow this. I think this is important because it allows some directories to be managed by Galaxy, some by infrastructure/VM, and some by instance (or any other break down). Maybe grep through the code for `galaxy.util.config_directories_from_setting`.

jmchilton commented:
Second note - documentation and expectations.

I spend more and more time thinking about the interface or surface area Galaxy exposes. Can we do this - but spend a year with the plugin system in beta. Can we specify in the admin documentation that plugins  are beta will need to be adjusted from release to release until the interface stabilizes and is formalized.

And like with the viz things - it would be my suggestion to add interesting plugins that aren't overly specific to a given instance to this repository instead of an external one. So that we can all see the interface Galaxy exposes and watch it grow and see how it is used.

Overall I'm really excited. While both great changes - I feel like in some ways switching to a JavaScript build system and the future change of encouraging pip installs of Galaxy and/or docker-galaxy-stable really have detrimental impacts on the fundamental "hackability" or "customizability" of Galaxy - this is a good stab in the opposite direction. I like it a lot in the abstract. Thanks a bunch.

hexylena commented:
@jmchilton predicted both of those first two concerns when I was chatting with @bgruening before this PR was opened :) 

I'll reply to one whilst he's asleep. For "just require the plugin to be present", @bgruening offered the comment that there are some plugins he only wants to be activated during an event of some sort, and thus was going to have multiple levels of activation, presence/absence and then a switch in an admin interface somewhere.

mvdbeek commented on doc/source/admin/webhooks.rst:
```diff
@@ -0,0 +1,187 @@
+
+Plugin activation
+-----------------
+Webhooks are deactivate by default.
```
s/deactivate/deactivated/

bgruening replied:
fixed, thanks.

bgruening committed:
tense changes

bgruening commented:
> - [ ] These operate different from IEs and Vizs that just require the plugin be present. Can we adjust the behavior to just load it if it is there? Maybe create a demo plugin directory and real plugin directory (the latter being empty) - and just update the documentation to state that the demo plugins can be enabled by adding that directory to the config.

I will create a demo directory, thats a good idea. Do you want the XKCD and PhD as a demo or as real? :)
Do you want me to remove the entire `activate: True` config option or is ok if all plugin have it as `activate: False` by default, which emulates the behavior of IEs and normal Vis. I thought to integrate this additional layer is a good idea because it might be useful to activate a few plugins only during workshops. Moreover, we have plans to allow activation/deactivation by a user and this infrastructure was the first step. But I'm happy to remove it as this is needed - no problem. 

> - [ ] Can we update the config option to allow multiple plugin directories. Tours, viz, and IEs allow this. I think this is important because it allows some directories to be managed by Galaxy, some by infrastructure/VM, and some by instance (or any other break down). Maybe grep through the code for `galaxy.util.config_directories_from_setting`.

I will do.

bgruening committed:
Change according to Aysam's review

bgruening committed:
Change according to Aysam's review

guerler commented on client/galaxy/scripts/layout/menu.js:
```diff
@@ -1,396 +1,418 @@
+            Webhooks.add({
+                url: 'api/webhooks/masthead/all',
+                callback: function(webhooks) {
+                    $(document).ready(function() {
```
Is this necessary?

anatskiy replied:
@guerler if you're talking about `$(document).ready()`, then it is necessary because after plugins have been retrieved, they are being added to `Galaxy.page.masthead`, which is available only after the full page load. Thus, I'm waiting for the full page load and only then add plugins to masthead.

guerler commented on client/galaxy/scripts/layout/menu.js:
```diff
@@ -1,396 +1,418 @@
+                                    url     : webhook.config.url,
+                                    tooltip : webhook.config.tooltip,
+                                    onclick : (typeof webhook.config.function != 'undefined') ? new Function(webhook.config.function) : '',
+                                    // visible : webhook.activate
```
Can be removed.

guerler commented on client/galaxy/scripts/layout/menu.js:
```diff
@@ -1,396 +1,418 @@
+                                    icon    : webhook.config.icon,
+                                    url     : webhook.config.url,
+                                    tooltip : webhook.config.tooltip,
+                                    onclick : (typeof webhook.config.function != 'undefined') ? new Function(webhook.config.function) : '',
```
onclick: webhook.config.function && (new Function(webhook.config.function))?

bgruening committed:
remove obsolete code

bgruening committed:
Merge branch 'dev' into feature/plugin-system

bgruening committed:
code optimization, thanks Sam

bgruening committed:
move all plugins into a demo folder

bgruening committed:
allow multiple webhook dirs

anatskiy committed:
Merge branch 'feature/plugin-system' of github.com:bgruening/galaxy into feature/plugin-system

anatskiy committed:
PEP8

anatskiy committed:
remove blank line

bgruening committed:
use config_directories_from_setting

bgruening committed:
merge

bgruening committed:
activate all plugins in the demo folder

anatskiy committed:
activate all demo plugins in testing purposes

bgruening committed:
Adopt help to the latest changes

bgruening committed:
add case where no user is logged in

bgruening committed:
Merge branch 'feature/plugin-system' of https://github.com/bgruening/galaxy into feature/plugin-system

bgruening committed:
Merge branch 'dev' into feature/plugin-system

bgruening commented:
@guerler we have addressed all your comments.
@jmchilton let me know if this is ok for you or if we should remove the `activate: true` option entirely. The current behavior should be the same as IE and tours and the additional level is just a bonus.

Thanks everyone for your kind words! Please let me know what else is needed :)
P.S. Any help with the merge/rebase problem is very welcome!

anatskiy committed:
Merge branch 'feature/plugin-system' of github.com:bgruening/galaxy into feature/plugin-system

anatskiy committed:
pep8

bgruening committed:
change default settings in galaxy.ini.sample

bgruening committed:
Add note about interface stability.

jmchilton commented:
@bgruening Are you saying they are `activate: true` by default? That is fine if it is a just a switch to optionally disable them. It's be cool if you updated viz and ies to have a similar switch then.

anatskiy committed:
fix failing Travis tests #2 (TOX_ENV=py27-lint) and #6 (TOX_ENV=py27-lint-imports)

bgruening commented:
@jmchilton correct, all components should be `true` with this commit: https://github.com/galaxyproject/galaxy/pull/3040/commits/f4dc38a54da3539f4d63f7e33708bcc2e33012fb

dannon commented:
Merged locally to resolve client build conflicts.

bgruening commented:
Thanks a lot Dannon!

nsoranzo commented:
This has somehow broken qunit tests, can somebody take a look? https://travis-ci.org/galaxyproject/galaxy/jobs/169936672

dannon commented:
@nsoranzo Sure, taking a look now.

dannon commented:
To follow up here, there was a bad merge of the client javascript files (and others) in the middle of this large PR.  Combined with the last push not triggering one of the tests (showing 3/3 green), this bad merge slipped through the cracks when the tests would have caught it.

I'm restoring the JS from the previous version and re-applying the changes linking in the webhooks changes (In a new PR)

martenson commented:
Bad merge within this PR caused https://github.com/galaxyproject/galaxy/pull/3203

martenson commented on client/galaxy/scripts/layout/menu.js:
```diff
@@ -1,396 +1,417 @@
-            title           : '',
-            url             : '',
-            target          : '_parent',
-            noscratchbook   : false
```
removal of this parameter does not seem correct

martenson replied:
ping @guerler - what do you think?

guerler replied:
You are right, but is this an outdated change?

martenson replied:
I see, it was remedied in https://github.com/galaxyproject/galaxy/pull/3102
Thanks!
